kong/pdk/client.lua

```diff
@@ -241,6 +242,50 @@ local function new(self)
     return is_tls and "tls" or "tcp"
   end
 
+  ---
+  -- Returns a table of `authenticated_groups` of the currently authenticated
+  -- consumer.
+  -- @function kong.client.get_authenticated_groups
+  -- @phases access, header_filter, body_filter, log
+  -- @treturn table,nil|err
+  -- The table will have an array part to iterate over, and a hash part where each
+  -- group name is indexed by itself. Eg.
+  -- {
+  --   [1] = "users",
+  --   [2] = "admins",
+  --   users = "users",
+  --   admins = "admins",
+  -- }
+  -- @usage
+  -- local groups, err = kong.client.get_authenticated_groups()
+  -- if not err then
+  --   for i = 1, #groups_to_check do
+  --     if groups[i] == 'group1' then
+  --       return true
+  --     end
+  --   end
+  -- end
+  function _CLIENT.get_authenticated_groups()
+    check_phase(AUTH_AND_LATER)
+
+    local authenticated_groups = ngx.ctx.authenticated_groups
+    if authenticated_groups == nil then
+      return nil
+    end
+
+    if type(authenticated_groups) ~= "table" then
+      return nil, "invalid authenticated_groups, a table was expected"
+    end
+
+    local groups = {}
+    for i = 1, #authenticated_groups do
```

Comment:
The PDK generally chooses safe over fast (its target audience is not the typical OpenResty programmer, but someone potentially new to both Kong and Lua). It's not a hard rule, but here I don't know if we want to allow the user to modify the set of authenticated groups by fiddling with the table returned by the getter.

Reply:
so our own plugins should not use the PDK then?

Reply:
@Tieske I'm afraid I don't follow. The implementation of `get_authenticated_groups` that is currently in the released version of `kong.plugins.acl.groups` already returns a copy.

Reply:
that comment wasn't based on the code, but on your comment:

> The PDK generally chooses safe over fast (its target audience is not the typical OpenResty programmer, but someone potentially new to both Kong and Lua)

If that's what it is, then we shouldn't be using it, right?

Reply:
> The implementation of get_authenticated_groups that is currently in the released version of kong.plugins.acl.groups already returns a copy.

The original didn't: https://github.com/Kong/kong-plugin-enterprise-canary/blob/master/kong/plugins/canary/groups.lua#L65

As commented [here](https://github.com/Kong/kong/pull/4013/files#r247037159) and [here](https://github.com/Kong/kong/pull/4932#discussion_r316240061) the whole idea was to have a single set of methods in the PDK. Initial steps were taken, but then reverted by changing the implementation in ACL (diverted from Canary again), instead of moving it into the PDK.

Now we're moving yet another implementation into the PDK???

Can we please do it right this time, and consolidate the code and move it into the PDK. Or if PDK is only for end users, move the base-code into Kong and add safe-wrapper of those methods in the PDK.

Reply:
> If that's what it is, then we shouldn't be using it, right?

No, we _should_ be using it, because reading the code from our plugins is how users learn how to write their own. Part of the work on introducing the PDK was porting over our plugins that used raw OpenResty calls to the PDK.

Reply:
Even if we decide to return a copy, we should use the [`table.clone`](https://github.com/openresty/luajit2#tableclone) method from OpenResty LuaJIT to make it faster and shorter than doing a manual copy.

In my view if we want more of our code to use the PDK functions we should make it at least possible to not incur additional overheads for all the callers. I'm okay keeping it "safe" by default, but having an optional argument like `no_copy` would make it much more suitable for core code's consumption (especially those that has already been accessing the group object directly and doesn't need a copy).

Reply:
I may be wrong, but I believe the explicit loop there was for adding both array and hash-style access to the resulting table (not sure if this is kept in the original, otherwise, sure, table.clone is a good choice — we should probably get rid of `kong.tools.utils.shallow_copy` as well).

I'm also okay with providing non-default fast-paths with reduced safety in the PDK if those prove necessary, but I'm just not sure if it's worth it here, because as far as I can see core code doesn't actually use authenticated groups, it's something in the realm of plugins only.

Reply:
I had a quick look around and couldn't find PDK functions that returns copies. Even getting credentials and consumers returns the real thing. 
Disclaimer: I didn't check everything.

If we return copies, we should also implement that in other places to be consistent across the board. If some return copies and others not, that's even more cumbersome.

Reply:
Agreed. @thibaultcha what's your view on this?